### DISCUSSION TITLE
PSA: DotNetCoreVerbosity enum has changed into a class - lolwat!

### DISCUSSION THREAD

**Discussion by frankradocaj** [cake-build/cake#3788] [General]
This is just a comment to help any poor developer who hits this issue in the future. I spent 3 hours debugging this, hours I will never get back! 😭

### Problem
The `DotNetCoreVerbosity` *enum* in [Cake v1.0](https://github.com/cake-build/cake/blob/v1.0.0/src/Cake.Common/Tools/DotNetCore/DotNetCoreVerbosity.cs) has changed into a *class* in [Cake v2.0](https://github.com/cake-build/cake/blob/v2.0.0/src/Cake.Common/Tools/DotNetCore/DotNetCoreVerbosity.cs). Diabolical.

If you are using any enum helper utils (I had code like `Enum.TryParse(...)`), you'll see nonsensical errors when you upgrade.

### Solution
Switch over to using the `DotNetVerbosity` enum, which mercifully is still an actual, you know, enum.

### Note
I only encountered this issue with this enum, but without looking through the whole codebase I suspect there are other enums that have been... err _classified_.

**Comment by augustoproiete** [comment]
👋🏻  This was by design in order to maintain compile-time compatibility with Cake v1 code, with the introduction of the new `DotNetVerbosity` in v2. Changing it to a class allowed us to [implicitly convert](https://github.com/cake-build/cake/blob/v2.0.0/src/Cake.Common/Tools/DotNetCore/DotNetCoreVerbosity.cs#L71-L95) `DotNetCoreVerbosity` to `DotNetVerbosity`  in all kinds of ways that C# allows us to, including equality comparisons, explicit casts, conversions, etc...

Two enums were changed to classes in v2: [`DotNetCoreVerbosity`](https://github.com/cake-build/cake/blob/v2.0.0/src/Cake.Common/Tools/DotNetCore/DotNetCoreVerbosity.cs) and [`DotNetCoreRollForward`](https://github.com/cake-build/cake/blob/v2.0.0/src/Cake.Common/Tools/DotNetCore/DotNetCoreRollForward.cs).

You seem to be using reflection to parse a string to an enum, so understandably you get a runtime error.

If you were using the "Cake way" of parsing arguments, you would not have been affected by this - as we included a converter.

```cakescript
var dotnetVerbosity = Argument<DotNetCoreVerbosity>("dotnetVerbosity", DotNetCoreVerbosity.Normal);
Information("dotnet verbosity set to {0}", dotnetVerbosity);
```

```powershell
dotnet cake build.cake --dotnetVerbosity=Diagnostic
```

**Reply by frankradocaj** [reply]
Oh, I didn't realise you could have an Argument parse an enum like that, nice! Thanks for responding. 🙏🏻
